File: dsa-interview/cpp/SegmentTree.cpp
/**
 * Segment Tree Implementation in C++
 * 
 * Time Complexity:
 * - Build: O(n)
 * - Query: O(log n)
 * - Update: O(log n)
 * 
 * Space Complexity: O(n)
 */

#include <iostream>
#include <vector>
#include <climits>
#include <functional>

// ==================== Generic Segment Tree ====================
template<typename T>
class SegmentTree {
private:
    std::vector<T> tree;
    std::vector<T> lazy;
    int n;
    T identity;
    std::function<T(T, T)> combine;
    
    void build(const std::vector<T>& nums, int node, int start, int end) {
        if (start == end) {
            tree[node] = nums[start];
        } else {
            int mid = start + (end - start) / 2;
            build(nums, 2 * node + 1, start, mid);
            build(nums, 2 * node + 2, mid + 1, end);
            tree[node] = combine(tree[2 * node + 1], tree[2 * node + 2]);
        }
    }
    
    void updateHelper(int node, int start, int end, int idx, T val) {
        if (start == end) {
            tree[node] = val;
        } else {
            int mid = start + (end - start) / 2;
            if (idx <= mid) {
                updateHelper(2 * node + 1, start, mid, idx, val);
            } else {
                updateHelper(2 * node + 2, mid + 1, end, idx, val);
            }
            tree[node] = combine(tree[2 * node + 1], tree[2 * node + 2]);
        }
    }
    
    T queryHelper(int node, int start, int end, int l, int r) {
        if (r < start || end < l) return identity;
        if (l <= start && end <= r) return tree[node];
        
        int mid = start + (end - start) / 2;
        return combine(
            queryHelper(2 * node + 1, start, mid, l, r),
            queryHelper(2 * node + 2, mid + 1, end, l, r)
        );
    }

public:
    SegmentTree(const std::vector<T>& nums, T identity, std::function<T(T, T)> combine)
        : n(nums.size()), identity(identity), combine(combine) {
        tree.resize(4 * n);
        lazy.resize(4 * n, identity);
        build(nums, 0, 0, n - 1);
    }
    
    void update(int idx, T val) {
        updateHelper(0, 0, n - 1, idx, val);
    }
    
    T query(int l, int r) {
        return queryHelper(0, 0, n - 1, l, r);
    }
};

// ==================== Segment Tree with Lazy Propagation ====================
class LazySegmentTree {
private:
    std::vector<long long> tree;
    std::vector<long long> lazy;
    int n;
    
    void pushDown(int node, int start, int end) {
        if (lazy[node] != 0) {
            tree[node] += (end - start + 1) * lazy[node];
            if (start != end) {
                lazy[2 * node + 1] += lazy[node];
                lazy[2 * node + 2] += lazy[node];
            }
            lazy[node] = 0;
        }
    }
    
    void build(const std::vector<int>& nums, int node, int start, int end) {
        if (start == end) {
            tree[node] = nums[start];
        } else {
            int mid = start + (end - start) / 2;
            build(nums, 2 * node + 1, start, mid);
            build(nums, 2 * node + 2, mid + 1, end);
            tree[node] = tree[2 * node + 1] + tree[2 * node + 2];
        }
    }
    
    void rangeUpdateHelper(int node, int start, int end, int l, int r, long long val) {
        pushDown(node, start, end);
        
        if (r < start || end < l) return;
        
        if (l <= start && end <= r) {
            tree[node] += (end - start + 1) * val;
            if (start != end) {
                lazy[2 * node + 1] += val;
                lazy[2 * node + 2] += val;
            }
            return;
        }
        
        int mid = start + (end - start) / 2;
        rangeUpdateHelper(2 * node + 1, start, mid, l, r, val);
        rangeUpdateHelper(2 * node + 2, mid + 1, end, l, r, val);
        tree[node] = tree[2 * node + 1] + tree[2 * node + 2];
    }
    
    long long queryHelper(int node, int start, int end, int l, int r) {
        pushDown(node, start, end);
        
        if (r < start || end < l) return 0;
        if (l <= start && end <= r) return tree[node];
        
        int mid = start + (end - start) / 2;
        return queryHelper(2 * node + 1, start, mid, l, r) +
               queryHelper(2 * node + 2, mid + 1, end, l, r);
    }

public:
    LazySegmentTree(const std::vector<int>& nums) : n(nums.size()) {
        tree.resize(4 * n, 0);
        lazy.resize(4 * n, 0);
        build(nums, 0, 0, n - 1);
    }
    
    void rangeUpdate(int l, int r, long long val) {
        rangeUpdateHelper(0, 0, n - 1, l, r, val);
    }
    
    long long query(int l, int r) {
        return queryHelper(0, 0, n - 1, l, r);
    }
};

// ==================== Merge Sort Tree (for K-th smallest in range) ====================
class MergeSortTree {
private:
    std::vector<std::vector<int>> tree;
    int n;
    
    void build(const std::vector<int>& nums, int node, int start, int end) {
        if (start == end) {
            tree[node] = {nums[start]};
        } else {
            int mid = start + (end - start) / 2;
            build(nums, 2 * node + 1, start, mid);
            build(nums, 2 * node + 2, mid + 1, end);
            
            // Merge sorted arrays
            std::merge(tree[2 * node + 1].begin(), tree[2 * node + 1].end(),
                      tree[2 * node + 2].begin(), tree[2 * node + 2].end(),
                      std::back_inserter(tree[node]));
        }
    }
    
    // Count elements <= val in range [l, r]
    int countLessOrEqual(int node, int start, int end, int l, int r, int val) {
        if (r < start || end < l) return 0;
        if (l <= start && end <= r) {
            return std::upper_bound(tree[node].begin(), tree[node].end(), val) - tree[node].begin();
        }
        
        int mid = start + (end - start) / 2;
        return countLessOrEqual(2 * node + 1, start, mid, l, r, val) +
               countLessOrEqual(2 * node + 2, mid + 1, end, l, r, val);
    }

public:
    MergeSortTree(const std::vector<int>& nums) : n(nums.size()) {
        tree.resize(4 * n);
        build(nums, 0, 0, n - 1);
    }
    
    // Find k-th smallest element in range [l, r] (1-indexed k)
    int kthSmallest(int l, int r, int k) {
        int lo = INT_MIN, hi = INT_MAX;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (countLessOrEqual(0, 0, n - 1, l, r, mid) < k) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        return lo;
    }
    
    int countInRange(int l, int r, int minVal, int maxVal) {
        return countLessOrEqual(0, 0, n - 1, l, r, maxVal) - 
               countLessOrEqual(0, 0, n - 1, l, r, minVal - 1);
    }
};

int main() {
    // Sum Segment Tree
    std::vector<int> nums = {1, 3, 5, 7, 9, 11};
    SegmentTree<int> sumTree(nums, 0, [](int a, int b) { return a + b; });
    
    std::cout << "Sum of range [1, 3]: " << sumTree.query(1, 3) << std::endl; // 15
    std::cout << "Sum of range [0, 5]: " << sumTree.query(0, 5) << std::endl; // 36
    
    sumTree.update(1, 10);
    std::cout << "After update, sum of range [1, 3]: " << sumTree.query(1, 3) << std::endl; // 22
    
    // Min Segment Tree
    SegmentTree<int> minTree(nums, INT_MAX, [](int a, int b) { return std::min(a, b); });
    std::cout << "Min in range [0, 3]: " << minTree.query(0, 3) << std::endl;
    
    // Lazy Segment Tree
    std::cout << "\n--- Lazy Segment Tree ---" << std::endl;
    LazySegmentTree lazyTree({1, 2, 3, 4, 5});
    std::cout << "Initial sum [0, 4]: " << lazyTree.query(0, 4) << std::endl; // 15
    lazyTree.rangeUpdate(1, 3, 10);  // Add 10 to indices 1, 2, 3
    std::cout << "After +10 to [1,3], sum [0, 4]: " << lazyTree.query(0, 4) << std::endl; // 45
    
    // Merge Sort Tree
    std::cout << "\n--- Merge Sort Tree ---" << std::endl;
    MergeSortTree mst({3, 1, 4, 1, 5, 9, 2, 6});
    std::cout << "2nd smallest in [0, 4]: " << mst.kthSmallest(0, 4, 2) << std::endl; // 1
    std::cout << "3rd smallest in [2, 6]: " << mst.kthSmallest(2, 6, 3) << std::endl; // 4
    
    return 0;
}

File: dsa-interview/java/SegmentTree.java
package java;

/**
 * Segment Tree Implementation
 * 
 * Time Complexity:
 * - Build: O(n)
 * - Query: O(log n)
 * - Update (point): O(log n)
 * - Update (range with lazy): O(log n)
 * 
 * Space Complexity: O(n)
 * 
 * Common Interview Problems:
 * - Range Sum Query - Mutable (Leetcode 307)
 * - Range Minimum Query
 * - Count of Smaller Numbers After Self (Leetcode 315)
 * - Range Sum Query 2D - Mutable (Leetcode 308)
 * - My Calendar I/II/III (Leetcode 729, 731, 732)
 */
public class SegmentTree {
    
    private int[] tree;
    private int[] lazy;  // For lazy propagation
    private int n;
    
    public SegmentTree(int[] nums) {
        if (nums == null || nums.length == 0) return;
        
        n = nums.length;
        tree = new int[4 * n];  // 4*n to be safe
        lazy = new int[4 * n];
        build(nums, 0, 0, n - 1);
    }
    
    private void build(int[] nums, int node, int start, int end) {
        if (start == end) {
            tree[node] = nums[start];
        } else {
            int mid = start + (end - start) / 2;
            int leftChild = 2 * node + 1;
            int rightChild = 2 * node + 2;
            
            build(nums, leftChild, start, mid);
            build(nums, rightChild, mid + 1, end);
            
            tree[node] = tree[leftChild] + tree[rightChild];  // Sum query
        }
    }
    
    /**
     * Point update: Update value at index idx to val.
     */
    public void update(int idx, int val) {
        update(0, 0, n - 1, idx, val);
    }
    
    private void update(int node, int start, int end, int idx, int val) {
        if (start == end) {
            tree[node] = val;
        } else {
            int mid = start + (end - start) / 2;
            int leftChild = 2 * node + 1;
            int rightChild = 2 * node + 2;
            
            if (idx <= mid) {
                update(leftChild, start, mid, idx, val);
            } else {
                update(rightChild, mid + 1, end, idx, val);
            }
            
            tree[node] = tree[leftChild] + tree[rightChild];
        }
    }
    
    /**
     * Range sum query from l to r (inclusive).
     */
    public int query(int l, int r) {
        return query(0, 0, n - 1, l, r);
    }
    
    private int query(int node, int start, int end, int l, int r) {
        if (r < start || end < l) {
            return 0;  // Out of range
        }
        
        if (l <= start && end <= r) {
            return tree[node];  // Completely in range
        }
        
        int mid = start + (end - start) / 2;
        int leftChild = 2 * node + 1;
        int rightChild = 2 * node + 2;
        
        int leftSum = query(leftChild, start, mid, l, r);
        int rightSum = query(rightChild, mid + 1, end, l, r);
        
        return leftSum + rightSum;
    }
    
    // ==================== Range Update with Lazy Propagation ====================
    
    /**
     * Range update: Add val to all elements in range [l, r].
     */
    public void rangeUpdate(int l, int r, int val) {
        rangeUpdate(0, 0, n - 1, l, r, val);
    }
    
    private void rangeUpdate(int node, int start, int end, int l, int r, int val) {
        // Push down any pending lazy updates
        pushDown(node, start, end);
        
        if (r < start || end < l) {
            return;  // Out of range
        }
        
        if (l <= start && end <= r) {
            // Completely in range - apply update lazily
            tree[node] += (end - start + 1) * val;
            if (start != end) {
                lazy[2 * node + 1] += val;
                lazy[2 * node + 2] += val;
            }
            return;
        }
        
        int mid = start + (end - start) / 2;
        int leftChild = 2 * node + 1;
        int rightChild = 2 * node + 2;
        
        rangeUpdate(leftChild, start, mid, l, r, val);
        rangeUpdate(rightChild, mid + 1, end, l, r, val);
        
        tree[node] = tree[leftChild] + tree[rightChild];
    }
    
    private void pushDown(int node, int start, int end) {
        if (lazy[node] != 0) {
            tree[node] += (end - start + 1) * lazy[node];
            if (start != end) {
                lazy[2 * node + 1] += lazy[node];
                lazy[2 * node + 2] += lazy[node];
            }
            lazy[node] = 0;
        }
    }
    
    /**
     * Query with lazy propagation.
     */
    public int queryLazy(int l, int r) {
        return queryLazy(0, 0, n - 1, l, r);
    }
    
    private int queryLazy(int node, int start, int end, int l, int r) {
        pushDown(node, start, end);
        
        if (r < start || end < l) {
            return 0;
        }
        
        if (l <= start && end <= r) {
            return tree[node];
        }
        
        int mid = start + (end - start) / 2;
        return queryLazy(2 * node + 1, start, mid, l, r) + 
               queryLazy(2 * node + 2, mid + 1, end, l, r);
    }
    
    // ==================== DEMO ====================
    public static void main(String[] args) {
        int[] nums = {1, 3, 5, 7, 9, 11};
        SegmentTree st = new SegmentTree(nums);
        
        // Sum queries
        System.out.println("Sum of range [1, 3]: " + st.query(1, 3)); // 3+5+7 = 15
        System.out.println("Sum of range [0, 5]: " + st.query(0, 5)); // 1+3+5+7+9+11 = 36
        
        // Point update
        st.update(1, 10);  // Change index 1 from 3 to 10
        System.out.println("After update, sum of range [1, 3]: " + st.query(1, 3)); // 10+5+7 = 22
        
        // Range Minimum Query example
        System.out.println("\n--- Range Minimum Query ---");
        RMQSegmentTree rmq = new RMQSegmentTree(new int[]{2, 5, 1, 4, 9, 3});
        System.out.println("Min in range [1, 4]: " + rmq.query(1, 4)); // 1
        System.out.println("Min in range [3, 5]: " + rmq.query(3, 5)); // 3
    }
}

/**
 * Range Minimum Query Segment Tree
 */
class RMQSegmentTree {
    private int[] tree;
    private int n;
    
    public RMQSegmentTree(int[] nums) {
        n = nums.length;
        tree = new int[4 * n];
        java.util.Arrays.fill(tree, Integer.MAX_VALUE);
        build(nums, 0, 0, n - 1);
    }
    
    private void build(int[] nums, int node, int start, int end) {
        if (start == end) {
            tree[node] = nums[start];
        } else {
            int mid = start + (end - start) / 2;
            build(nums, 2 * node + 1, start, mid);
            build(nums, 2 * node + 2, mid + 1, end);
            tree[node] = Math.min(tree[2 * node + 1], tree[2 * node + 2]);
        }
    }
    
    public int query(int l, int r) {
        return query(0, 0, n - 1, l, r);
    }
    
    private int query(int node, int start, int end, int l, int r) {
        if (r < start || end < l) return Integer.MAX_VALUE;
        if (l <= start && end <= r) return tree[node];
        
        int mid = start + (end - start) / 2;
        return Math.min(
            query(2 * node + 1, start, mid, l, r),
            query(2 * node + 2, mid + 1, end, l, r)
        );
    }
    
    public void update(int idx, int val) {
        update(0, 0, n - 1, idx, val);
    }
    
    private void update(int node, int start, int end, int idx, int val) {
        if (start == end) {
            tree[node] = val;
        } else {
            int mid = start + (end - start) / 2;
            if (idx <= mid) {
                update(2 * node + 1, start, mid, idx, val);
            } else {
                update(2 * node + 2, mid + 1, end, idx, val);
            }
            tree[node] = Math.min(tree[2 * node + 1], tree[2 * node + 2]);
        }
    }
}
